### PR TITLE
[KYUUBI #3098] Unify the event log code path

### PR DESCRIPTION
**Pull request by zhaomin1423**
### _Why are the changes needed?_

close https://github.com/apache/incubator-kyuubi/issues/3098#issue-1309741627
### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by ulysses-you** [comment]
thank you @zhaomin1423 , can you put some details about this pr ? it can help reviewer a lot

**Comment by zhaomin1423** [comment]
We provide spark, json, jdbc and custom in the option of kyuubi.engine.event.loggers, but only implement spark and json, the pr implement the custom, it can add custom logger by putting the jar in the classpath.

File: kyuubi-events/src/main/scala/org/apache/kyuubi/events/EventHandlerRegister.scala [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            .foreach(EventBus.register)
+
+        case other =>
+          warn(s"Unsupported event logger: ${other.toString}")
+      }
+  }
+
+  protected def createSparkEventHandler(kyuubiConf: KyuubiConf)
+      : Option[EventHandler[KyuubiEvent]] = {
+    Option.empty
```
Comment (ulysses-you) [inline kyuubi-events/src/main/scala/org/apache/kyuubi/events/EventHandlerRegister.scala:63]: we should throw exception if child deos not implemet the corresponding method

File: kyuubi-events/src/main/scala/org/apache/kyuubi/events/EventHandlerRegister.scala [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        case EventLoggerType.JDBC =>
+          loadJdbcEventHandler(conf)
+            .foreach(EventBus.register)
+
+        case EventLoggerType.CUSTOM =>
+          EventHandlerLoader.loadCustom(conf)
+            .foreach(EventBus.register)
+
+        case other =>
+          warn(s"Unsupported event logger: ${other.toString}")
```
Comment (ulysses-you) [inline kyuubi-events/src/main/scala/org/apache/kyuubi/events/EventHandlerRegister.scala:57]: throw exception ?

File: kyuubi-events/src/main/scala/org/apache/kyuubi/events/EventHandlerRegister.scala [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      : Option[EventHandler[KyuubiEvent]] = {
+    Option.empty
+  }
+
+  protected def createJdbcEventHandler(kyuubiConf: KyuubiConf)
+      : Option[EventHandler[KyuubiEvent]] = {
+    Option.empty
+  }
+
+  private def loadSparkEventHandler(kyuubiConf: KyuubiConf): Option[EventHandler[KyuubiEvent]] = {
```
Comment (ulysses-you) [inline kyuubi-events/src/main/scala/org/apache/kyuubi/events/EventHandlerRegister.scala:76]: can we add a generic method to support arbitrary event handler ? so we do not need each method for spark/json/jdbc etc..

File: kyuubi-events/src/main/scala/org/apache/kyuubi/events/handler/CustomEventHandlerProvider.scala [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kyuubi.events.handler
+
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.events.KyuubiEvent
+
+trait CustomEventHandlerProvider {
```
Comment (ulysses-you) [inline kyuubi-events/src/main/scala/org/apache/kyuubi/events/handler/CustomEventHandlerProvider.scala:22]: this is a new feature, we can do it in a new pr. so the main change of this pr is: unify the event log code path.

**Comment by zhaomin1423** [comment]
Thanks for your review. Loading custom event logger will do in a new pr, others have been updated. @ulysses-you

File: kyuubi-events/src/main/scala/org/apache/kyuubi/events/EventHandlerRegister.scala [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    loggers
+      .map(EventLoggerType.withName)
+      .foreach {
+        loadEventHandler(_, conf)
+          .foreach(EventBus.register)
+      }
+  }
+
+  protected def createSparkEventHandler(kyuubiConf: KyuubiConf)
+      : Option[EventHandler[KyuubiEvent]] = {
```
Comment (ulysses-you) [inline kyuubi-events/src/main/scala/org/apache/kyuubi/events/EventHandlerRegister.scala:47]: seems option is unnecessary for these methods ? if we do not support then we will throw exception

Comment (ulysses-you) [reply kyuubi-events/src/main/scala/org/apache/kyuubi/events/EventHandlerRegister.scala:47]: then we can inline these methods

File: kyuubi-events/src/main/scala/org/apache/kyuubi/events/EventHandlerRegister.scala [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+import org.apache.kyuubi.events.handler.EventHandler
+
+trait EventHandlerRegister extends Logging {
+
+  def registerEngineEventLoggers(conf: KyuubiConf): Unit = {
+    val loggers = conf.get(ENGINE_EVENT_LOGGERS)
+    register(loggers, conf)
+  }
+
+  def registerServiceEventLoggers(conf: KyuubiConf): Unit = {
```
Comment (ulysses-you) [inline kyuubi-events/src/main/scala/org/apache/kyuubi/events/EventHandlerRegister.scala:32]: `Service` -> `Server`

**Review by ulysses-you — COMMENTED** [review]
@zhaomin1423 sorry for the late, thank you for the work

File: externals/kyuubi-hive-sql-engine/src/main/scala/org/apache/kyuubi/engine/hive/events/HiveEventHandlerRegister.scala [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  override protected def createJsonEventHandler(
+      kyuubiConf: KyuubiConf): EventHandler[KyuubiEvent] = {
+    val hadoopConf = KyuubiHadoopUtils.newHadoopConf(kyuubiConf)
+    val hostName = InetAddress.getLocalHost.getCanonicalHostName
+    val handler = HiveJsonLoggingEventHandler(
+      s"Hive-$hostName",
+      ENGINE_EVENT_JSON_LOG_PATH,
+      hadoopConf,
+      kyuubiConf)
+    handler
```
Comment (ulysses-you) [inline externals/kyuubi-hive-sql-engine/src/main/scala/org/apache/kyuubi/engine/hive/events/HiveEventHandlerRegister.scala:39]: nit: unnecessary variable

Comment (ulysses-you) [reply externals/kyuubi-hive-sql-engine/src/main/scala/org/apache/kyuubi/engine/hive/events/HiveEventHandlerRegister.scala:39]: for all similar code path

**Review by ulysses-you — APPROVED** [review]

File: externals/kyuubi-hive-sql-engine/src/main/scala/org/apache/kyuubi/engine/hive/events/HiveEventHandlerRegister.scala (hunk 7 above)
Comment (zhaomin1423) [reply externals/kyuubi-hive-sql-engine/src/main/scala/org/apache/kyuubi/engine/hive/events/HiveEventHandlerRegister.scala]: Thanks, updated.

**Comment by ulysses-you** [comment]
thanks, merging to master

**Pull request closed without merging**